File: backend/app/api/v1/routers/workouts.py
"""Public workout generator endpoint (rate-limited)."""

from typing import Annotated

from fastapi import APIRouter, Depends, Request

from app.api.v1.deps import provide_generate_workout
from app.api.v1.schemas.workout import WorkoutGenerateRequest, WorkoutItemRead, WorkoutRead
from app.application.dto.workout import WorkoutRequest
from app.application.use_cases.workout_use_cases import GenerateWorkout
from app.core.rate_limit import RATE_LIMIT, limiter

router = APIRouter(prefix="/workouts", tags=["workouts"])


@router.post(
    "/generate",
    response_model=WorkoutRead,
    summary="Generate a basic routine from a goal and minimal attributes",
)
@limiter.limit(RATE_LIMIT)
async def generate_workout(
    request: Request,  # required by slowapi to identify the client
    payload: WorkoutGenerateRequest,
    use_case: Annotated[GenerateWorkout, Depends(provide_generate_workout)],
) -> WorkoutRead:
    result = await use_case.execute(
        WorkoutRequest(
            goal=payload.goal,
            experience=payload.experience,
            height_cm=payload.height_cm,
            weight_kg=payload.weight_kg,
            sex=payload.sex,
            age=payload.age,
            equipment=payload.equipment,
        )
    )
    template = result.template
    return WorkoutRead(
        goal=template.goal,
        name=template.name,
        description=template.description,
        difficulty=template.difficulty,
        bmi=result.bmi,
        bmi_category=result.bmi_category,
        items=[WorkoutItemRead.model_validate(item) for item in template.items],
    )

File: frontend/src/api/workouts.ts
// Workout generator API call. Attributes go in the POST body (not the query
// string) so personal data stays out of URLs.

import { api } from './client'
import { toExercise, type ExercisePayload } from './explorer'
import type { Difficulty, Equipment, Goal, Workout } from './types'
import { i18n } from '@/i18n'

export interface GenerateRequest {
  goal: Goal
  experience: Difficulty
  heightCm: number
  weightKg: number
  sex?: 'male' | 'female' | 'other'
  age?: number
  equipment?: Equipment[]
}

interface WorkoutItemPayload {
  exercise: ExercisePayload
  sets: number
  reps: string
  rest_seconds: number
}

interface WorkoutPayload {
  goal: Goal
  name: string
  description: string
  difficulty: Difficulty
  bmi: number
  bmi_category: string
  items: WorkoutItemPayload[]
}

export async function generateWorkout(req: GenerateRequest): Promise<Workout> {
  const lang = i18n.global.locale.value
  const payload = await api.post<WorkoutPayload>(`/workouts/generate?lang=${lang}`, {
    goal: req.goal,
    experience: req.experience,
    height_cm: req.heightCm,
    weight_kg: req.weightKg,
    sex: req.sex,
    age: req.age,
    equipment: req.equipment,
  })
  return {
    goal: payload.goal,
    name: payload.name,
    description: payload.description,
    difficulty: payload.difficulty,
    bmi: payload.bmi,
    bmiCategory: payload.bmi_category,
    items: payload.items.map((i) => ({
      exercise: toExercise(i.exercise),
      sets: i.sets,
      reps: i.reps,
      restSeconds: i.rest_seconds,
    })),
  }
}
